src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -0,0 +1,255 @@
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentException(nameof(output));
+
+            Expand(hashAlgorithmName, hashLength, prk, output, info);
+        }
+
+        private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            if (prk.Length < hashLength)
+                throw new ArgumentException(nameof(prk));
+
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(output));
+
+            Span<byte> counter = stackalloc byte[1];
+            Span<byte> t = Span<byte>.Empty;
+            Span<byte> remainingOutput = output;
+
+            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+            {
+                for (int i = 1; ; i++)
+                {
+                    hmac.AppendData(t);
+                    hmac.AppendData(info);
+                    counter[0] = (byte)i;
+                    hmac.AppendData(counter);
+
+                    if (remainingOutput.Length >= hashLength)
+                    {
+                        t = remainingOutput.Slice(0, hashLength);
+                        remainingOutput = remainingOutput.Slice(hashLength);
+                        GetHashAndReset(hmac, t);
+                    }
+                    else
+                    {
+                        if (remainingOutput.Length > 0)
+                        {
+                            Span<byte> lastChunk = stackalloc byte[hashLength];
+                            GetHashAndReset(hmac, lastChunk);
+                            lastChunk.Slice(0, remainingOutput.Length).CopyTo(remainingOutput);
+                        }
+
+                        break;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Performs the key derivation HKDF Expand and Extract functions
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="ikm">Input keying material</param>
+        /// <param name="outputLength">Length of the output keying material</param>
+        /// <param name="salt">Optional salt value (a non-secret random value). If not provided it defaults to a byte array of <see cref="HashLength"/> zeros.</param>
+        /// <param name="info">Optional context and application specific information</param>
+        /// <returns>Output keying material</returns>
+        public static byte[] DeriveKey(HashAlgorithmName hashAlgorithmName, byte[] ikm, int outputLength, byte[] salt = null, byte[] info = null)
+        {
+            if (ikm == null)
+                throw new ArgumentNullException(nameof(ikm));
+
+            int hashLength = HashLength(hashAlgorithmName);
+            Span<byte> prk = stackalloc byte[hashLength];
```

Comment:
I assume we're guaranteed `hashLength` is going to be small?

Reply:
64 at most for now (SHA512)